File: online-training-platform/leetcode/symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    bool isSymmetricHelper(TreeNode* first, TreeNode* second) {
        
        if(first == nullptr && second == nullptr) {
            return true;
        }
        
        if(first != nullptr && second == nullptr){
            return false;
        }
        
        if(first == nullptr && second != nullptr){
            return false;
        }
        
        if(first->val != second->val) {
            return false;
        }
        
        return isSymmetricHelper(first->right, second->left) && 
            isSymmetricHelper(first->left, second->right);
    }
    
    bool isSymmetric(TreeNode* root) {
        if(root->left != nullptr && root->right == nullptr){
            return false;
        }
        
        if(root->left == nullptr && root->right != nullptr){
            return false;
        }
        
        return isSymmetricHelper(root->left, root->right);
    }
};

File: online-training-platform/leetcode/symmetric-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    
    fun isSymmetricHelper(left: TreeNode?, right: TreeNode?): Boolean{
        
        if(left == null && right == null){
            return true
        }
        
        if(left != null && right == null){
            return false
        }
        
        if(left == null && right != null) {
            return false
        }
        
        if(left?.`val` != right?.`val`){
            return false
        }
        
        return isSymmetricHelper(left?.right, right?.left) && isSymmetricHelper(left?.left, right?.right)
    }
    
    fun isSymmetric(root: TreeNode?): Boolean {
        if(root == null){
            return true
        }
        
        return isSymmetricHelper(root.left, root.right)
    }
}